### ISSUE TITLE
Generating Folder Hashes on large file trees slows down Contao

### ISSUE THREAD

**Issue by funkjunkee** [contao/core#8798]
https://github.com/contao/core/blob/2474ff9a8e0cf0f533818461d20930e49baaa9b0/system/modules/core/library/Contao/Folder.php#L372

We've observed through stacktrace monitoring (New Relic), that when a file is uploaded in the backend, then generating/updating the hash for the parent folder does take a very long time (several minutes per file) IF there's a lot of files scattered around in multiple folders witin the path of the file. According to our monitioring, this is due to the fact, that each file name found within the path (RecursiveIteratorIterator operation) is concattenated into a very long string, for which an md5 hash is generated. Going through the filesystem recursively is basically what's killing performance in our case.

I've tried to research if the \Folder::getHash function is used in any way to check a folders integrity, but wasn't able to find anything such, thus, I'm deceived to believe, that this function may not be of any good use after all. So, in case the hash is not used for checking the integrity of the filesystem, then wouldn't it make sense to skip recursive directory traversal and generate the hash ONLY for the path to the folder?

**Comment by leofeyer** [comment]
> I've tried to research if the \Folder::getHash function is used in any way to check a folders integrity, but wasn't able to find anything such, thus, I'm deceived to believe, that this function may not be of any good use after all.

The hash is used to find folders moved e.g. via FTP when synchronizing the filesystem with the database.

> According to our monitioring, this is due to the fact, that each file name found within the path (RecursiveIteratorIterator operation) is concattenated into a very long string, for which an md5 hash is generated. Going through the filesystem recursively is basically what's killing performance in our case.

@contao/developers Can we optimize this? Maybe hash the file paths so the concatenated string does not become too long?

**Comment by ausi** [comment]
Couldn’t the hash of a folder just be a hash of all children’s hashes combined?

**Comment by Toflar** [comment]
Sounds like a blockchain 😄  Should work imho but might still take long when you have a lot of subfolders. But it should work if we only do it for the topmost level of each folder and start from the very bottom so hashes are updated bottom to top.

**Comment by ausi** [comment]
like so?

- folder, hash(`abc` + `def` + `mno`) = `pqr`
  - file1, hash `abc`
  - file2, hash `def`
  - subfolder, hash(`ghi` + `jkl`) = `mno`
     - file1, hash `ghi`
     - file2, hash `jkl`

**Comment by Toflar** [comment]
Exactly.

**Comment by funkjunkee** [comment]
The main problem in our experience is, that all files included in all folders are used in the hash as well. If it's only used to determine changes in the folder structure, then the getHash function could be reduced to only include the subpath of each folder in the tree and skip any non-dir files.

**Comment by ausi** [comment]
I think hashing the hashes of files combined has the exact same features as currently but should be way faster... subfolders don’t have to be handled differently, only the hash of the subfolder has to be used. Hashing only paths wouldn’t change if the contents of a file change.

**Comment by funkjunkee** [comment]
Here's some monitoring examples to show what happens.

Below is an image which show the total amount of time used by various functions and methods in the trace.
![image](https://user-images.githubusercontent.com/25210400/31442778-c7caad54-ae97-11e7-8300-f91fad366d8f.png)

As you can see, Contao\Folder::getHash and Contao\Folder::__get are using the majority of time within this trace.

Drilling down into the stack trace, you can see how getHash is executed in the sequence:
![image](https://user-images.githubusercontent.com/25210400/31442990-632cbfe4-ae98-11e7-8e75-75c3e5cc1719.png)

In this example, it takes about 94seconds to update the hash of 1 folder only, in the whole trace, 4 folders are updated.

**Comment by funkjunkee** [comment]
I've beefed up our PHP tracing and was recording the files and directories included in the process. Here's a quick depiction of what happens.

Sample Folder Structure:

- dir1
     - subdir1
     - subdir2
     - subdir3
          - sub-subdir_A
          - sub-subdir_B
          - **sub-subdir_C**
               - file-1
               - file-2
               - **file-3**

When uploading a file to **sub-subdir_C**, then the getHash function will include all files included within all subdirectoris within dir1, which in our case will easily include thousands of files.

This essentially slows down the performance of the File manager to a point, where it takes several minutes to uplod a single 50kB file.

**Comment by funkjunkee** [comment]
Hi All,
while this is still work in progress, is there any idea how we could temporarily make this faster? Our users are struggling more and more each day.

Would it be feasible, to change the function to only create an md5 hash of the foldername without recursing the through the whole file tree? We don't have any FTP access enabled.

**Comment by ausi** [comment]
You can try to replace the `getHash()` method in https://github.com/contao/core/blob/3e41ad8718f6dea9c10c8a9b0341ed86f2bbe2ae/system/modules/core/library/Contao/Folder.php#L372 with:

```php
protected function getHash()
{
	return md5($this->strFolder);
}
```

This will have the downside that renaming the folder via FTP can no longer be detected and it may cause other issues as well.

**Comment by fritzmg** [comment]
May be this should be an optional feature in general? None of our clients (and myself included) ever utilize it.

**Comment by ausi** [comment]
I don’t think we have to make it optional. The solution as described in https://github.com/contao/core/issues/8798#issuecomment-335801932 should work very fast without problems. I’m working on it, but my time is limited. (Can be increased via <https://www.patreon.com/ausi> ;)

**Comment by ausi** [comment]
@funkjunkee could you please check if the changes from #8856 fix the issue in your installation?

**Comment by ausi** [comment]
@funkjunkee The changes from #8856 are now released in Version 3.5.34.

It would be great if you could update your installation and give some feedback on the performance of the new folder hashing method.
